lib/queuePopulator/IngestionReader.js

```diff
@@ -57,14 +60,115 @@ class IngestionReader extends LogReader {
+        let statusResponse;
+        async.mapSeries(initPathNodes, (pathNode, cb) => {
+            // if complete, we do not care to get other zookeeper data
+            if (statusResponse === 'complete') {
+                return process.nextTick(cb);
+            }
+            const path = `${this.bucketInitPath}/${pathNode}`;
+            return this.zkClient.getData(path, (err, data) => {
+                if (err) {
+                    if (err.name !== 'NO_NODE') {
+                        logger.error(
+                            'Could not fetch ingestion init state',
+                            { method: 'IngestionReader._readInitState',
+                              zkPath: path,
+                              error: err });
+                        return cb(err);
+                    }
+                    return this.zkClient.mkdirp(path, err => {
+                        if (err) {
+                            logger.error(
+                                'Could not pre-create path in zookeeper',
+                                { method: 'IngestionReader._readInitState',
+                                  zkPath: path,
+                                  error: err });
+                            return cb(err);
+                        }
+                        return cb();
+                    });
+                }
+                logger.debug('fetched ingestion init state node', {
+                    method: 'IngestionReader._readInitState',
+                    zkPath: path,
+                });
+                const d = data && data.toString();
+                if (pathNode === 'status') {
+                    statusResponse = d;
+                }
+                return cb(null, d);
+            });
+        }, (err, data) => {
+            if (err) {
+                return done(err);
+            }
+            const [status, versionMarker, keyMarker] = data;
+            return done(null, {
+                status: status || 'incomplete',
+                versionMarker,
+                keyMarker
+            });
+        });
+    }
+
+    /**
+     * Set the init (snapshot) state for this given IngestionReader
+     * @param {object} initState - initState (snapshot) for ingestion
+     * @param {string} initState.status - complete || incomplete
+     * @param {string} [initState.versionMarker] - NextVersionIdMarker
+     * @param {string} [initState.keyMarker] - NextKeyMarker
+     * @param {Logger.newRequestLogger} logger - request logger object
+     * @param {function} done - callback(error)
+     * @return {undefined}
+     */
+    _writeInitState(initState, logger, done) {
+        // initState is set by each request of processLogEntries. If undefined,
+        // we did not go through snapshot phase
+        if (!initState) {
+            return process.nextTick(done);
+        }
+        const initPathNodes = ['status', 'versionMarker', 'keyMarker'];
```

Comment:
Is it not a good idea to use a constant because it's an array?

Reply:
It's good but if it's being repeated, it's better to declare them once and share them across. This makes it easy to keep track of them instead of having to hunt for them in the code.